File: Projects/Web.RIA/app/Krista.FM.RIA.Extensions.Marks.OMSU/Presentation/Controls/IneffExpensesControl.js
ViewPersistence.isDirty = function () {
    return false;
}

ViewPersistence.refresh = function () {
    targetMarkStore.reload()
}

ShowDifferenceWarning = function () {
    var warningsFound = false;
    for (i = 0; i < targetFactsStore.getCount(); i++) {
        warningsFound = warningsFound || (targetFactsStore.getAt(i).get('HasWarning') != '');
    }
    if (warningsFound) {
        Ext.Msg.show({
            title: 'Внимание!',
            msg: 'В некоторых МО есть устаревшие значения показателя!',
            buttons: Ext.Msg.OK,
            animEl: 'targetFactsGrid',
            icon: Ext.MessageBox.EXCLAMATION
        });
    }
}

FactsBackToEditHandler = function () {
    var s = targetFactsGrid.getSelectionModel().getSelections();
    for (i = 0; i < s.length; i++) {
        s[i].set('RefStatusData', 1);
    }
    targetFactsGrid.save();
}

FactsApproveHandler = function () {
    var s = targetFactsGrid.getSelectionModel().getSelections();
    for (i = 0; i < s.length; i++) {
        if (s[i].get('RefStatusData') < 3) {
            s[i].set('RefStatusData', 3);
        }
    }
    targetFactsStore.save();
}

TooltipShowHandler = function () {
    var rowIndex = targetFactsGrid.view.findRowIndex(this.triggerElement);
    var cellIndex = targetFactsGrid.view.findCellIndex(this.triggerElement);
    var record = targetFactsStore.getAt(rowIndex);

    if (!record) {
        this.hide();
        return false;
    }

    var fieldName = targetFactsGrid.getColumnModel().getDataIndex(cellIndex);
    var data = record.get(fieldName);

    if (cellIndex == 1) {
        if (data == 1) { this.body.dom.innerHTML = 'На редактировании'; }
        else if (data == 2) { this.body.dom.innerHTML = 'На рассмотрении'; }
        else if (data == 3) { this.body.dom.innerHTML = 'Утвержден'; }
        else if (data == 4) { this.body.dom.innerHTML = 'Принят'; }
        else {
            this.body.dom.innerHTML = '';
            this.hide();
            return false; 
        }
    }
    else if ((cellIndex == 2) && (data != '')) {
        this.body.dom.innerHTML = 'Утвержденное и рассчитанное значения не совпадают';
    }
    else {
        this.body.dom.innerHTML = '';
        this.hide();
        return false;
    }
}

RefStatusDataRenderer = function (value, p, record) {    
    if (value != null) {
        var status = record.get('RefStatusData');
        if (status == 1) { return '<img class="x-panel-inline-icon icon-useredit" src="/extjs/resources/images/default/s-gif/ext.axd" alt="">'; }
        else if (status == 2) { return '<img class="x-panel-inline-icon icon-usermagnify" src="/extjs/resources/images/default/s-gif/ext.axd" alt="">'; }
        else if (status == 3) { return '<img class="x-panel-inline-icon icon-usertick" src="/extjs/resources/images/default/s-gif/ext.axd" alt="">'; }
        else if (status == 4) { return '<img class="x-panel-inline-icon icon-accept" src="/extjs/resources/images/default/s-gif/ext.axd" alt="">'; }
    } 
}

HasWarningRenderer = function (value, p, record) {
    if (record.get('HasWarning') != '') {
        return '<img class="x-panel-inline-icon icon-exclamation" src="/extjs/resources/images/default/s-gif/ext.axd" alt="!">';
    }
}

TargetFactPriorRenderer = function (value, p, record) {
    return ValueRenderer(value, targetMarkPrecision.getValue(), record.get('HasWarnPrior') != '');
}

TargetFactCurrentRenderer = function (value, p, record) {
    return ValueRenderer(value, targetMarkPrecision.getValue(), record.get('HasWarnCurrent') != '');
}

ValueRenderer = function (value, precision, redHighlight) {
    var result = FormatUsing(value, precision);
    if (redHighlight) {
        return '<span class="red-text-cell">' + result + '</span>'
    }
    return result;
}

SrcFactsNameRenderer = function (value, p, record) {
    var indent = record.get('Level') * 30;
    return '<div style="display:inline-block; padding-left:' + indent + 'px">' + value + '</span>'
}

SrcFactValueRenderer = function (value, p, record) {
    return ValueRenderer(value, record.get('Precision'), false);
}

TargetMarkStoreLoadHandler = function () {
    targetMarkForm.getForm().loadRecord(targetMarkStore.getAt(0));
    targetMarkDescriptionLabel.setText(targetMarkDescription.getValue(), true);
    targetMarkCalcMarkLabel.setText(targetMarkCalcMark.getValue(), true);
    targetMarkMeasureUnitLabel.setText(targetMarkMeasureUnit.getValue(), true);

    targetFactsStore.reload();
    targetFactsGrid.getColumnModel().getColumnById('PriorCalc').header = (currentYear.getValue() - 1) + ' рассчитано';
    targetFactsGrid.getColumnModel().getColumnById('PriorApproved').header = (currentYear.getValue() - 1) + ' утверждено';
    targetFactsGrid.getColumnModel().getColumnById('CurrentCalc').header = currentYear.getValue() + ' рассчитано';
    targetFactsGrid.getColumnModel().getColumnById('CurrentApproved').header = currentYear.getValue() + ' утверждено';

    sourceFactsGrid.getColumnModel().getColumnById('srcFactsColPrior').header = (currentYear.getValue() - 1);
    sourceFactsGrid.getColumnModel().getColumnById('srcFactsColCurrent').header = currentYear.getValue();
    sourceFactsGrid.setDisabled(true);
}

function FormatUsing(value, precision) {
    var usePrecision = 0;
    if (precision) {
        usePrecision = precision;
    }
    formatStr = '0.000,' + String.leftPad('', usePrecision, '0');
    result = Ext.util.Format.number(value, formatStr + '/i');
    return result;
}

File: Projects/Web.RIA/app/Krista.FM.RIA.Extensions.Marks.OMSU/Presentation/Controls/IneffExpensesControl.cs
using System;
using System.Collections.Generic;
using System.Web.Mvc;
using Ext.Net;
using Krista.FM.Extensions;
using Krista.FM.RIA.Core;
using Krista.FM.RIA.Core.Principal;

namespace Krista.FM.RIA.Extensions.MarksOMSU.Presentation.Controls
{
    public class IneffExpensesControl
    {
        private readonly string itfCaptionShort;
        private readonly string targetMarkGetMethodName;

        public IneffExpensesControl(string targetMarkGetMethodName, string itfCaption)
        {
            this.itfCaptionShort = itfCaption;
            this.targetMarkGetMethodName = targetMarkGetMethodName;
        }

        public List<Component> Build(ViewPage page)
        {
            var resourceManager = ResourceManager.GetInstance(page);
            resourceManager.RegisterScript("ViewPersistence", "/Content/js/ViewPersistence.js");
            resourceManager.RegisterScript("Hack", "/Content/js/Ext.util.Format.number.Hack.js");
            resourceManager.RegisterClientScriptBlock("IneffExpensesControl", Resources.Resource.IneffExpensesControl);

            if (!ExtNet.IsAjaxRequest)
            {
                resourceManager.RegisterClientStyleBlock("CustomStyle", @".red-text-cell { color: #FF0000 !important; }");
                resourceManager.RegisterIcon(Icon.UserEdit);
                resourceManager.RegisterIcon(Icon.UserMagnify);
                resourceManager.RegisterIcon(Icon.UserTick);
                resourceManager.RegisterIcon(Icon.Accept);
                resourceManager.RegisterIcon(Icon.Exclamation);
            }

            InstallTargetMarkStore(page);

            InstallTargetFactsStore(page);

            InstallSourceFactsStore(page);

            var viewport =
                new Viewport
                    {
                        ID = "viewportMain",
                        Items =
                            {
                                new BorderLayout
                                    {
                                        ID = "borderLayoutMain",
                                        North = { Items = { CreateTargetMarkPanel() } },
                                        Center = { Items = { CreateTargetFactsPanel() } },
                                        South = { Items = { CreateSourceFactsPanel(page) } }
                                    }
                            }
                    };

            return new List<Component> { viewport, CreateTargetFactsTooltip() };
        }

        private static void InstallTargetFactsStore(ViewPage page)
        {
            var targetFactsStore =
                new Store
                    {
                        AutoLoad = false,
                        ID = "targetFactsStore",
                        Proxy =
                            {
                                new HttpProxy
                                    {
                                        Url = "/IneffExpenses/GetTargetFacts", 
                                        Method = HttpMethod.POST
                                    }
                            },
                        UpdateProxy =
                            {
                                new HttpWriteProxy
                                    {
                                        Url = "/IneffExpenses/SaveTargetFacts", 
                                        Method = HttpMethod.POST, 
                                        Timeout = 500000
                                    }
                            },
                        BaseParams =
                        { 
                            new Parameter
                                {
                                    Name = "targetMarkId", 
                                    Value = "targetMarkID.getValue()", 
                                    Mode = ParameterMode.Raw
                                } 
                        },
                        WarningOnDirty = false,
                        Reader =
                            {
                                new JsonReader
                                    {
                                        IDProperty = "ID", 
                                        Root = "data", 
                                        Fields =
                                            {
                                                new RecordField("ID"), 
                                                new RecordField("RegionID"), 
                                                new RecordField("RefStatusData"),
                                                new RecordField("HasWarning"),
                                                new RecordField("NameMO"),

                                                new RecordField("HasWarnPrior"),
                                                new RecordField("PriorApproved"), 
                                                new RecordField("PriorCalc"),
                                                new RecordField("HasWarnCurrent"),
                                                new RecordField("CurrentApproved"), 
                                                new RecordField("CurrentCalc"),
                                            }
                                    }
                            },
                        Listeners =
                            {
                                LoadException = { Handler = "Ext.Msg.alert('Ошибка при загрузке списка значений целевых показателей', response.responseText);" },
                                Load = { Fn = "ShowDifferenceWarning" }
                            }
                    };
            page.Controls.Add(targetFactsStore);
        }

        private static void InstallSourceFactsStore(ViewPage page)
        {
            var sourceFactsStore =
                new Store
                    {
                        AutoLoad = false,
                        ID = "sourceFactsStore",
                        Proxy =
                            {
                                new HttpProxy
                                    {
                                        Url = "/IneffExpenses/GetSourceFacts", 
                                        Method = HttpMethod.POST
                                    }
                            },
                        BaseParams =
                        { 
                            new Parameter
                                {
                                    Name = "targetMarkId", 
                                    Value = "targetMarkID.getValue()", 
                                    Mode = ParameterMode.Raw
                                },
                            new Parameter
                                {
                                    Name = "regionId",
                                    Value = "targetFactsGrid.getSelectionModel().getSelected().get('RegionID')", 
                                    Mode = ParameterMode.Raw
                                } 
                        },
                        WarningOnDirty = false,
                        Reader =
                            {
                                new JsonReader
                                    {
                                        IDProperty = "ID", 
                                        Root = "data", 
                                        Fields =
                                            {
                                                new RecordField("ID"), 
                                                new RecordField("Symbol"),
                                                new RecordField("Formula"),
                                                new RecordField("Prior"),
                                                new RecordField("Current"),
                                                new RecordField("Name"),
                                                new RecordField("Level"),
                                                new RecordField("MeasureUnit"),
                                                new RecordField("Precision"),
                                            }
                                    }
                            },
                        Listeners =
                            {
                                LoadException = { Handler = "Ext.Msg.alert('Ошибка при загрузке списка значений исходных показателей', response.responseText);" },
                            }
                    };

            page.Controls.Add(sourceFactsStore);
        }

        private static Toolbar CreateTargetFactsToolbar()
        {
            var toolbar = new Toolbar();

            toolbar.Items.Add(
                new Button
                    {
                        ID = "expandDetails",
                        Icon = Icon.Help,
                        ToolTip = "Отобразить панель с исходными показателями",
                        Listeners = { Click = { Handler = "sourceFactsGrid.expand(true);" } }
                    });
           
            if (((BasePrincipal)System.Web.HttpContext.Current.User).IsInRole(MarksOMSUConstants.IneffGkhCalculateRole))
            {
                toolbar.Items.Add(new ToolbarSeparator());
                toolbar.Items.Add(
                    new Button
                        {
                            ID = "revertButton",
                            Icon = Icon.Pencil,
                            ToolTip = "Вернуть расчет показателя для выбранных МО на редактирование",
                            Listeners = { Click = { Fn = "FactsBackToEditHandler" } }
                        });
                toolbar.Items.Add(
                    new Button
                        {
                            ID = "approveButton",
                            Icon = Icon.Tick,
                            ToolTip = "Утвердить расчет показателя для выбранных МО",
                            Listeners = { Click = { Fn = "FactsApproveHandler" } }
                        });
            }

            return toolbar;
        }

        private static ToolTip CreateTargetFactsTooltip()
        {
            var toolTip = new ToolTip
            {
                ID = "RowTip",
                Target = "={targetFactsGrid.getView().mainBody}",
                Delegate = ".x-grid3-cell",
                TrackMouse = true,
                Listeners = { Show = { Fn = "TooltipShowHandler" } }
            };

            return toolTip;
        }

        private static IEnumerable<Component> CreateTargetFactsPanel()
        {
            var targetFactsGrid =
                new GridPanel
                    {
                        ID = "targetFactsGrid",
                        Title = "Значения итоговых показателей",
                        StoreID = "targetFactsStore",
                        Border = false,
                        MonitorResize = true,
                        AutoScroll = true,
                        ColumnLines = true,
                        LoadMask = { ShowMask = true, Msg = "Пересчет показателей..." },
                        SaveMask = { ShowMask = true, Msg = "Сохранение..." },

                        TopBar = { CreateTargetFactsToolbar() },

                        ColumnModel =
                            {
                                Columns =
                                    {
                                        new Column { ColumnID = "ID", DataIndex = "ID", Hidden = true, Fixed = true },
                                        new Column
                                            {
                                                ColumnID = "RefStatusData",
                                                DataIndex = "RefStatusData",
                                                Header = String.Empty,
                                                Width = 30,
                                                Fixed = true,
                                                Renderer = { Fn = "RefStatusDataRenderer" }
                                            },
                                        new Column
                                            {
                                                ColumnID = "HasWarning",
                                                DataIndex = "HasWarning",
                                                Header = String.Empty,
                                                Width = 30,
                                                Fixed = true,
                                                Renderer = { Fn = "HasWarningRenderer" }
                                            },
                                        new Column { ColumnID = "NameMO", DataIndex = "NameMO", Header = "Наименование МО", Width = 140 },                                        
                                        new Column { ColumnID = "HasWarnPrior", DataIndex = "HasWarnPrior", Header = String.Empty, Hidden = true },
                                        new Column
                                            {
                                                ColumnID = "PriorCalc", 
                                                DataIndex = "PriorCalc", 
                                                Header = "Рассчет",
                                                Css = "background-color: #EDEDED; text-align: right;", 
                                                Width = 85,
                                                Renderer = { Fn = "TargetFactPriorRenderer" }
                                            },
                                        new Column
                                            {
                                                ColumnID = "PriorApproved",
                                                DataIndex = "PriorApproved",
                                                Header = "Утверждено",
                                                Css = "background-color: #EDEDED; text-align: right;",
                                                Width = 85,
                                                Renderer = { Fn = "TargetFactPriorRenderer" }
                                            },
                                        new Column { ColumnID = "HasWarnCurrent", DataIndex = "HasWarnCurrent", Header = String.Empty, Hidden = true },
                                        new Column
                                            {
                                                ColumnID = "CurrentCalc", 
                                                DataIndex = "CurrentCalc", 
                                                Header = "Рассчет", 
                                                Width = 85,
                                                Css = "text-align: right;",
                                                Renderer = { Fn = "TargetFactCurrentRenderer" }
                                            },
                                        new Column
                                            {
                                                ColumnID = "CurrentApproved", 
                                                DataIndex = "CurrentApproved", 
                                                Header = "Утверждено", 
                                                Css = "text-align: right;",
                                                Width = 85,
                                                Renderer = { Fn = "TargetFactCurrentRenderer" }
                                            }
                                    }
                            },
                        SelectionModel =
                            {
                                new RowSelectionModel
                                    {
                                        SingleSelect = true,
                                        Listeners = { RowSelect = { Handler = @"sourceFactsStore.reload(); sourceFactsGrid.setDisabled(false);" } }
                                    }
                            },
                    };

            return new List<Component> { targetFactsGrid };
        }

        private static IEnumerable<Component> CreateSourceFactsPanel(ViewPage page)
        {
            var sourceFactsGrid =
                new GridPanel
                    {
                        ID = "sourceFactsGrid",
                        Title = "Значения исходных показателей по выбранному МО",
                        StoreID = "sourceFactsStore",
                        Border = false,
                        Height = 200,
                        Collapsed = true,
                        Collapsible = true,
                        Split = true,
                        MonitorResize = true,
                        AutoScroll = true,
                        ColumnLines = true,
                        LoadMask = { ShowMask = true, Msg = "Загрузка показателей..." },
                        Disabled = true,

                        ColumnModel =
                            {
                                Columns =
                                    {
                                        new Column { ColumnID = "srcFactsColID", DataIndex = "ID", Hidden = true },
                                        new Column { ColumnID = "srcFactsColLevel", DataIndex = "Level", Hidden = true },
                                        new Column
                                            {
                                                ColumnID = "srcFactsColName", 
                                                DataIndex = "Name", 
                                                Header = "Наименование", 
                                                Width = 400,
                                                Wrap = true,
                                                Renderer = { Fn = "SrcFactsNameRenderer" }
                                            },
                                        new Column { ColumnID = "srcFactsColSymbol", DataIndex = "Symbol", Header = "Обозначение", Width = 80 },
                                        new Column { ColumnID = "srcFactsColFormula", DataIndex = "Formula",  Header = "Формула", Width = 120 },
                                        new Column
                                            {
                                                ColumnID = "srcFactsColPrior", 
                                                DataIndex = "Prior", 
                                                Header = "(Загрузка...)", 
                                                Css = "background-color: #EDEDED; text-align: right;", 
                                                Width = 85,
                                                Renderer = { Fn = "SrcFactValueRenderer" }
                                            },
                                        new Column
                                            {
                                                ColumnID = "srcFactsColCurrent",
                                                DataIndex = "Current",
                                                Header = "(Загрузка...)",
                                                Css = "text-align: right;",
                                                Width = 85,
                                                Renderer = { Fn = "SrcFactValueRenderer" }
                                            },
                                        new Column { ColumnID = "srcFactsColMeasureUnit", DataIndex = "MeasureUnit",  Header = "Единица измерения", Width = 100 },
                                    }
                            },
                        SelectionModel = { new RowSelectionModel { SingleSelect = true } }
                    };

            sourceFactsGrid.AddColumnsWrapStylesToPage(page);
            return new List<Component> { sourceFactsGrid };
        }
        
        private IEnumerable<Component> CreateTargetMarkPanel()
        {
            var targetMarkPanel =
                new FormPanel
                    {
                        ID = "targetMarkForm",
                        Title = itfCaptionShort,
                        Border = false,
                        Padding = 5,
                        Layout = "form",
                        Collapsed = false,
                        Collapsible = true,
                        Height = 150,
                        LabelWidth = 130,
                        LabelAlign = LabelAlign.Right,
                        DefaultAnchor = "99%",
                        BodyCssClass = "x-window-mc",
                        CssClass = "x-window-mc",

                        TopBar =
                            {
                                new Toolbar
                                    {
                                        Items =
                                            {
                                                new Button
                                                    {
                                                        ID = "refreshButton",
                                                        Icon = Icon.PageRefresh,
                                                        ToolTip = "Обновить",
                                                        Listeners = { Click = { Handler = "ViewPersistence.refresh.call(window);" } }
                                                    },
                                                new Button
                                                    {
                                                        ID = "exportXlsButton",
                                                        Icon = Icon.PageExcel,
                                                        ToolTip = "Выгрузка в Excel",
                                                        DirectEvents =
                                                            {
                                                                Click =
                                                                    {
                                                                        Url = "/IneffExpenses/ExportTargetFactsToXls", 
                                                                        IsUpload = true, 
                                                                        CleanRequest = true,
                                                                        ExtraParams =
                                                                            {
                                                                                new Parameter("targetMarkId", "targetMarkID.getValue()", ParameterMode.Raw),
                                                                                new Parameter("itfCaption", "'{0}'".FormatWith(itfCaptionShort), ParameterMode.Raw),
                                                                            } 
                                                                    },
                                                            }
                                                    }
                                            }
                                    }
                            },

                        Items =
                            {
                                new Label { ID = "targetMarkDescriptionLabel", FieldLabel = "Описание", Text = "(Загрузка...)", LabelStyle = "font-weight:bold;" },
                                new Label { ID = "targetMarkCalcMarkLabel", FieldLabel = "Формула", Text = "(Загрузка...)", LabelStyle = "font-weight:bold;" },
                                new Label { ID = "targetMarkMeasureUnitLabel", FieldLabel = "Единица измерения", Text = "(Загрузка...)", LabelStyle = "font-weight:bold;" },
                                new TextField { ID = "targetMarkID", DataIndex = "ID", Hidden = true, ReadOnly = true },
                                new TextField { ID = "targetMarkDescription", DataIndex = "Description", Hidden = true, ReadOnly = true },
                                new TextField { ID = "targetMarkCalcMark", DataIndex = "CalcMark", Hidden = true, ReadOnly = true },
                                new TextField { ID = "currentYear", DataIndex = "Year", Hidden = true, ReadOnly = true },
                                new TextField { ID = "targetMarkMeasureUnit", DataIndex = "MeasureUnit", Hidden = true, ReadOnly = true },
                                new TextField { ID = "targetMarkPrecision", DataIndex = "Precision", Hidden = true, ReadOnly = true },
                            }
                    };
            return new List<Component> { targetMarkPanel };
        }
        
        private void InstallTargetMarkStore(ViewPage page)
        {
            var targetMarkStore =
                new Store
                    {
                        AutoLoad = true,
                        ID = "targetMarkStore",
                        Proxy =
                            {
                                new HttpProxy
                                    {
                                        Url = "/IneffExpenses/" + targetMarkGetMethodName,
                                        Method = HttpMethod.POST
                                    }
                            },
                        WarningOnDirty = false,
                        Reader =
                            {
                                new JsonReader
                                    {
                                        IDProperty = "ID", 
                                        Root = "data", 
                                        Fields =
                                            {
                                                new RecordField("ID"), 
                                                new RecordField("Name"),
                                                new RecordField("Description"),
                                                new RecordField("CalcMark"),
                                                new RecordField("Year"),
                                                new RecordField("MeasureUnit"),
                                                new RecordField("Precision"),
                                            }
                                    }
                            },
                        Listeners =
                            {
                                LoadException = { Handler = "Ext.Msg.alert('Ошибка при загрузке описания целевого показателя', response.responseText);" },
                                Load = { Handler = "TargetMarkStoreLoadHandler(); sourceFactsGrid.collapse(true);" }
                            }
                    };
            
            page.Controls.Add(targetMarkStore);        
        }
    }
}
